### ISSUE TITLE
Type parameter inference fails when it's only constrained by a wildcard

### ISSUE THREAD

**Issue by olafurpg** [lampepfl/dotty#2397]
The following code compiles with Scala 2.x 

```scala
package bar
import java.nio.file._
import java.util.stream.Collectors
object foo {
  Files.walk(Paths.get("")).collect(Collectors.toList[Path])
}
```

while dotc at commit 5093126421e333b2329399736e6d7ad327e3bdfa reports a type error

```
-- [E007] Type Mismatch Error: foo.scala:5:36 ----------------------------------
5 |  Files.walk(Paths.get("")).collect(Collectors.toList[Path])
  |                                    ^^^^^^^^^^^^^^^^^^^^^^^
  |       found:    java.util.stream.Collector[java.nio.file.Path, _,
  |         java.util.List[java.nio.file.Path]
  |       ]
  |       required: java.util.stream.Collector[_ >: java.nio.file.Path, A, R]
  |
  |       where:    A is a type variable
  |                 R is a type variable

one error found
```

**Comment by smarter** [comment]
You can make it work by using an intermediate variable:
```scala
import java.nio.file._
import java.util.stream.Collectors

object Test {
  val x = Collectors.toList[Path]
  Files.walk(Paths.get("")).collect(x)
}
```
The issue is with the second type parameter of `collect`, which is `x.A` here. We cannot give it a more precise type because in the return type of `toList` it's just a Java wildcard. If we don't use an intermediate variable then we just have no way of representing it. scalac represents it as `?0` but I don't know how that works under the hood.

**Comment by smarter** [comment]
I've managed to minimize it so that it doesn't depend on Java APIs, just wildcards:
```scala
class Foo[A]

object Test {
  def foo[T](x: Foo[T]) = x

  foo((new Foo[Int]: Foo[_]))
}
```
```scala
12 |  foo((new Foo[Int]: Foo[_]))
   |      ^^^^^^^^^^^^^^^^^^^^^^
   |      found:    Foo[_]
   |      required: Foo[T]
   |      
   |      where:    T is a type variable
```

**Comment by odersky** [comment]
It's fixed with #2402. scalac is different because there types with wildcards are existentials, and existentials are unpacked before usage. The `?0` you are seeing is a skolem constant arising from unpacking.
